src/scripts/background.js

```diff
@@ -250,18 +265,168 @@ function setDefaultValueForNewSettings() {
+        secEnd,
+        hourStart,
+        minStart,
+        secStart
+    }) || 0;
+    return {duration, startTime};
+}
+
+const postUserActivity = async (requestBody) => {
+    try {
+        await fetch(TRACK_USER_ACTIVITY_URL, {
+            method: 'POST',
+            headers: {
+                'Content-Type': 'application/json'
+            },
+            body: JSON.stringify(requestBody)
+        });
+    } catch(error) {
+        console.error(error);
+    }
+}
+
+const getDataFromStorage = (itemName, defaultValue) => {
+    return new Promise((resolve, reject) => {
+        storage.getValue(itemName, result => {
+            resolve(result || defaultValue);
+        });
+    });
+}
+
+const trackUserActivity = async () => {
+    const userEmail = await getDataFromStorage(STORAGE_USER_EMAIL, '');
+    const latitude = await getDataFromStorage(USER_LOCATION_LAT, null);
+    const longitude = await getDataFromStorage(USER_LOCATION_LONG, null);
+    let listItems = timeIntervalList || [];
+    listItems = listItems.filter(item => item.day === todayLocalDate());
+    const activityArray = listItems.map(item => {
+        const {duration, startTime} = mapTime(item);
+        return {
+            url: item.url,
+            duration,
+            startTime
+        };
+    });
+    const filteredActivity = activityArray.filter(item => item.duration);
+    if (filteredActivity.length) {
+        const requestBody =  {
+            user: userEmail,
+            location: {
+                latitude,
+                longitude
+            },
+            activity: filteredActivity
+        };
+        postUserActivity(requestBody);
+    }
+}
+
+const trackUserActivityHelper = async (lastActiveTabUrl) => {
+    const whiteList = await getDataFromStorage(STORAGE_WHITE_LIST, []);
+    const tabFromWhiteList = whiteList.find(item => lastActiveTabUrl.includes(item.split('://')[1]));
+    if (tabFromWhiteList) trackUserActivity();
+}
+
 function addListener() {
-    chrome.tabs.onActivated.addListener(function(info) {
-        chrome.tabs.get(info.tabId, function(tab) {
+    chrome.tabs.onActivated.addListener(activeInfo => {
+        chrome.tabs.get(activeInfo.tabId, async (tab) => {
             activity.addTab(tab);
+            await trackUserActivityHelper(lastActiveTabUrl);
+            lastActiveTabUrl = tab.url;
+            tabToUrl[activeInfo.tabId] = tab.url;
         });
     });
 
+    chrome.tabs.onUpdated.addListener(async (tabId, changeInfo, tab) => {
+        if (changeInfo.status === 'complete') {
+            if (lastActiveTabUrl !== tab.url) {
+                await trackUserActivityHelper(lastActiveTabUrl);
+            }
+            lastActiveTabUrl = tab.url;
+            tabToUrl[tabId] = tab.url;
+        }
+    });
+
+    chrome.tabs.onRemoved.addListener(async (tabId, removeInfo) => {
+        if (removeInfo.isWindowClosing) {
+            return;
+        }
+        if ( tabToUrl[tabId] !== lastActiveTabUrl) {
```

Comment:
redundant space?

Reply:
yep, didn't notice. I'll remove it with final commit